### PR TITLE
GRPCServer improvements

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit c86fd3eac71d24c3ca7fde36d151083f9b463241 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59730/c86fd3eac71d24c3ca7fde36d151083f9b463241/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59730/c86fd3eac71d24c3ca7fde36d151083f9b463241/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Server/GRPCServer.cpp [hunk 1]
```diff
@@ -484,19 +477,12 @@ namespace
...
         {
-            if (!query_info.has_value())
-            {
-                callback(false);
-                return;
-            }
-
-            query_info_ = std::move(query_info).value();
-            query_info.reset();
+            query_info_ = std::move(query_info);
```
Comment (jkartseva) [inline src/Server/GRPCServer.cpp:485]: After the `std::move` the state of `query_info` depends on the implementation of the move assignment operator of the third-party`grpc::QueryInfo` type.

Btw, C++ standard guaranties a valid but unspecified state of the moved-from object. This is why I believe std::optional was used in the first place.

File: src/Server/GRPCServer.cpp [hunk 2]
```diff
@@ -484,19 +477,12 @@ namespace
...
             callback(true);
```
Comment (jkartseva) [inline src/Server/GRPCServer.cpp:486]: For several consecutive calls of `GRPCServer:: read` the `callback` will always be called with `true`, even though the `query_info` is a moved-from object (presumably reset by the move assignment operator).

Is this the indented behavior?

File: programs/server/Server.cpp [hunk 3]
```diff
@@ -2380,6 +2380,16 @@ void Server::createServers(
         }
 
 #if USE_GRPC
+        using GRPCQueryInfo = clickhouse::grpc::QueryInfo;
+        GRPCQueryInfo local_query_info;
+        chassert(local_query_info.settings().size() == 0); /// NOLINT
```
Comment (jkartseva) [inline programs/server/Server.cpp:2385]: `empty()` ?

File: programs/server/Server.cpp [hunk 4]
```diff
@@ -2380,6 +2380,16 @@ void Server::createServers(
...
+
+        SettingsChanges settings_changes;
+        for (const auto & [key, value] : local_query_info.settings())
+            settings_changes.push_back({key, value});
```
Comment (jkartseva) [inline programs/server/Server.cpp:2389]: `emplace_back(key, value);` ?

**Review by jkartseva — COMMENTED** [review]
A test with consecutive calls of `GRPCServer:: read` would be great.

**Pull request closed without merging**